src/fontman/raster_font.cpp

```diff
@@ -0,0 +1,490 @@
+
+    if((w <= 0) || (h <= 0))
+    {
+        pLogWarning("Wrong width and height values! %d x %d",  w, h);
+        return;
+    }
+
+    if(!Files::fileExists(root + texFile))
+    {
+        pLogWarning("Failed to load font texture! file not exists: %s",
+                    (root + texFile).c_str());
+        return;
+    }
+
+    StdPicture fontTexture = XRender::LoadPicture(root + texFile);
+
+    if(!fontTexture.inited)
+        pLogWarning("Failed to load font texture! Invalid image!");
+
+    m_texturesBank.push_back(fontTexture);
+    StdPicture *loadedTexture = &m_texturesBank.back();
+
+    if((m_letterWidth == 0) || (m_letterHeight == 0))
+    {
+        m_letterWidth    = static_cast<uint32_t>(fontTexture.w) / w;
+        m_letterHeight   = static_cast<uint32_t>(fontTexture.h) / h;
+
+        if(m_spaceWidth == 0)
+            m_spaceWidth = m_letterWidth;
+
+        if(m_newlineOffset == 0)
+            m_newlineOffset = m_letterHeight;
+    }
+
+    font.endGroup();
+    font.beginGroup("entries");
+    std::vector<std::string> entries = font.allKeys();
+
+    for(std::string &x : entries)
+    {
+        std::string charPosX = "0", charPosY = "0";
+
+        std::string::size_type begPos = 0;
+        std::string::size_type endPos = x.find('-', begPos);
+
+        //QStringList tmp = x.split('-');
+        if(endPos == std::string::npos)
+            endPos = x.size();//Use entire string
+
+        charPosX = x.substr(begPos, endPos);
+
+        if(charPosX.empty())
+        {
+            pLogDebug("=invalid-X=%d=", x.c_str());
+            continue;
+        }
+
+        if(m_matrixWidth > 1)
+        {
+            if(endPos == x.size())
+            {
+                pLogWarning("=invalid-Y=%d= in the raster font map %s", x.c_str(), fontmap_ini.c_str());
+                continue;
+            }
+            begPos = endPos + 1;//+1 to skip '-' divider
+            endPos = x.find('-', begPos);
+            if(endPos == std::string::npos)
+                endPos = x.size();//Use entire string
+            charPosY = x.substr(begPos, endPos);
+            if(charPosY.empty())
+            {
+                pLogWarning("=invalid-Y=%d= in the raster font map %s", x.c_str(), fontmap_ini.c_str());
+                continue;
+            }
+        }
+
+        std::string charX;
+        font.read(x.c_str(), charX, "");
+
+        /*Format of entry: X23
+         * X - UTF-8 Symbol
+         * 2 - padding left [for non-mono fonts]
+         * 3 - padding right [for non-mono fonts]
+        */
+        if(charX.empty())
+            continue;
+
+        std::u32string ucharX = std_to_utf32(charX);
+        char32_t ch = ucharX[0];
+        //qDebug()<<"=char=" << ch << "=id="<<charPosX.toInt()<<charPosY.toInt()<<"=";
+        RasChar rch;
+#ifndef DEBUG_BUILD
+        try
+        {
+#endif
+            rch.tx              =  loadedTexture;
+            rch.l               =  std::stof(charPosY.c_str()) / m_matrixWidth;
```

Comment:
Better yet, I’d rather just use a grid, and ONLY store grid index (row * n_cola + col) as a uint16_t. A bit less flexible, but MUCH easier to store. I want to be able to support as many scripts as possible even on our tiniest platforms.

Reply:
Good point. I think we could certainly remove these t/b/l/r which seem to be intermediates leftover from Moondust, and if we want to minimize RAM we could also simply do grid-index (int16_t), width (uint8_t), and tex-index (uint8_t) which could also indicate valid/invalid.
I do care about this because if we want to support Latin, Cyrillic, and Kana (not even Hangul or Hanzi/Kanji/Hanja), then the current data structure alone would take up at least 8KB.

Reply:
Okay, I took a small look on this: I added these left/top/right/bottom to simplify don't re-compute glyph coordinates for the needs of OpenGL. I guess I can easily get rid of these fields with no harm.